### PR TITLE
Add SimpleNorm class

### PR DESCRIPTION
**Pull request by larrybradley**
### Description

This PR adds a `SimpleNorm` class that is analogous to the `simple_norm` function (I've been meaning to do this for a while).  The idea here is that `SimpleNorm` object can be applied to different images unlike the `simple_norm` function.  Functions that display images could also accept a `SimpleNorm` instance as a keyword.

One note -- the `simple_norm` keywords have been added incrementally and as such they are no longer grouped logically.  I've reordered the `SimpleNorm` keywords to be grouped logically.  I also reordered the interval keywords (percent, min/max_percent, v_min/max) in order of precedent (later keywords override earlier keywords).

Can I introduce a breaking change in this PR to `simple_norm` to reorder its keywords and require keyword-only arguments?

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
For ease of comparison...

New: 

```python
        stretch="linear",
        percent=None,
        min_percent=None,
        max_percent=None,
        vmin=None,
        vmax=None,
        power=1.0,
        log_a=1000,
        asinh_a=0.1,
        sinh_a=0.3,
        clip=False,
        invalid=-1.0,
```

Old:

```python
    stretch="linear",
    power=1.0,
    asinh_a=0.1,
    vmin=None,
    vmax=None,
    min_percent=None,
    max_percent=None,
    percent=None,
    clip=False,
    log_a=1000,
    invalid=-1.0,
    sinh_a=0.3,
```

The new order does make more sense but we are also only a few days till feature freeze... 🎲

**Comment by neutrinoceros** [comment]
> Can I introduce a breaking change in this PR to simple_norm to reorder its keywords and require keyword-only arguments?

*If* we choose to reorder them in one clean cut (as opposed to a deprecation cycle), we *should* absolutely make them keyword-only at the same time. Note that it is also possible to do it incrementally by emitting a warning for future-keyword-only arguments when they are passed positionally, and I've played around this idea in the past:
- here's a gist for doing this in pure Python https://gist.github.com/neutrinoceros/3f703d63761794b5e966330d51436b3c
- I also wrote a similar decorator as a C extension that lives in astropy https://github.com/astropy/astropy/blob/main/astropy/cosmology/_signature_deprecations.c  This one warns about future positional-only arguments, but unless we suspect this class is being instanciated in a hotloop somewhere, the small overhead from the pure-Python version is likely acceptable.

but admittedly this is a much more involved route, maintenance-wise.
I think we *could* roll the dice and go with your clean cut approach for the 7.0.0rc1 as long as we're pro-active in testing the release candidate downstream, and if we're ready to revert this change for the final release in case it breaks too many things, and then go the long road with an actual deprecation cycle starting in 7.1 (and ending not before 9.0).

**Comment by pllim** [comment]
> now is the time to do it

I don't see why we cannot do it later. I think early in the dev cycle of v8.0 would be nicer to users (if they bother to run tests against astropy dev wheels).

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/17217#issuecomment-2428418364

I vaguely remember we also discussed performance implications, besides extra maintenance.

**Comment by pllim** [comment]
Do we need What's New?

**Comment by larrybradley** [comment]
@pllim I added a What's New entry.

Let's just get this in and possibly address the `simple_norm` for a later release.

**Comment by neutrinoceros** [comment]
> Let's just get this in and possibly address the simple_norm keyword order for a later release.

I'll open an issue for it tomorrow unless one of you wishes to beat me to it, and I vow to address it early in the 7.1 cycle with a deprecation.

File: astropy/visualization/mpl_normalize.py [hunk 1]
```diff
@@ -212,6 +212,231 @@ def inverse(self, values, invalid=None):
...
+        norm = snorm(image)
+        fig, ax = plt.subplots()
+        axim = ax.imshow(image, norm=norm, origin='lower')
+        fig.colorbar(axim)
+    """
+
+    def __init__(
+        self,
+        stretch="linear",
+        percent=None,
```
Comment (neutrinoceros) [inline astropy/visualization/mpl_normalize.py:312]: Let's make every argument beyond `stretch` keyword-only from the get go
```suggestion
        *,
        percent=None,
```

File: astropy/visualization/mpl_normalize.py [hunk 2]
```diff
@@ -212,6 +212,231 @@ def inverse(self, values, invalid=None):
...
+        sinh_a=0.3,
+        clip=False,
+        invalid=-1.0,
+    ):
+        if percent is not None:
+            interval = PercentileInterval(percent)
+        elif min_percent is not None or max_percent is not None:
+            interval = AsymmetricPercentileInterval(
+                min_percent or 0.0, max_percent or 100.0
+            )
```
Comment (neutrinoceros) [inline astropy/visualization/mpl_normalize.py:329]: Hum, `max_percent or 100.0` will silently replace an explicit `max_percent=0.0` with `100.0`, which I think would be surprising and could lead to subtle and hard to debug situations.
Is there a good reason for `AsymmetricPercentileInterval.__init__` not to have default values for `min_percent` and `max_percent` ?
How about the following

```suggestion
            interval = AsymmetricPercentileInterval(
                lower_percentile=min_percent, upper_percentile=max_percent
            )
```

together with this patch:
```patch
diff --git a/astropy/visualization/interval.py b/astropy/visualization/interval.py
index 66fac6a117..b370da0152 100644
--- a/astropy/visualization/interval.py
+++ b/astropy/visualization/interval.py
@@ -156,9 +156,9 @@ class AsymmetricPercentileInterval(BaseInterval):
         sampled from the array (with replacement).
     """

-    def __init__(self, lower_percentile, upper_percentile, n_samples=None):
-        self.lower_percentile = lower_percentile
-        self.upper_percentile = upper_percentile
+    def __init__(self, lower_percentile=None, upper_percentile=None, n_samples=None):
+        self.lower_percentile = lower_percentile if lower_percentile is not None else 0.0
+        self.upper_percentile = upper_percentile if upper_percentile is not None else 100.0
         self.n_samples = n_samples

     def get_limits(self, values):
```

File: astropy/visualization/tests/test_norm.py [hunk 3]
```diff
@@ -281,6 +286,40 @@ def test_invalid_stretch(self):
...
+    data = np.arange(25).reshape((5, 5))
+    fig, ax = plt.subplots()
+    snorm = SimpleNorm("sqrt", percent=99)
+    axim = snorm.imshow(data, ax=ax)
+    assert isinstance(axim, AxesImage)
+    keys = ("vmin", "vmax", "stretch", "clip", "invalid")
+    for key in keys:
+        assert getattr(axim.norm, key) == getattr(snorm(data), key)
+
+    plt.clf()
```
Comment (neutrinoceros) [inline astropy/visualization/tests/test_norm.py:312]: `clf` is discouraged (though not formally deprecated)
https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.FigureBase.clf

```suggestion
    fig.clear()
```

File: astropy/visualization/mpl_normalize.py [hunk 4]
```diff
@@ -212,6 +212,231 @@ def inverse(self, values, invalid=None):
...
+            axim = snorm.imshow(image, ax=ax, origin='lower')
+            fig.colorbar(axim)
+        """
+        import matplotlib.pyplot as plt
+
+        if ax is None:
+            ax = plt.gca()
+
+        if "X" in kwargs:
+            raise ValueError("Cannot give both ``X`` and ``data``")
```
Comment (neutrinoceros) [inline astropy/visualization/mpl_normalize.py:428]: do we really need this branch ? I would expect `Axis.imshow` to handle this case itself.

**Review by neutrinoceros — COMMENTED** [review]
Here's a couple comments and suggestions. I would also suggest adding a parametrized test to cover norms other than `"sqrt"` and `"linear"`, even if the test doesn't do much more than running through the init routine.

File: astropy/visualization/mpl_normalize.py (hunk 1 above)
Comment (larrybradley) [reply astropy/visualization/mpl_normalize.py:312]: I'd prefer to make it keyword-only after `percent`.  99.99% of the time I only use `percent` interactively and that would save me a lot of typing.  :smile: And I use this functionality a lot!

Addressing your other comments now, thanks!

File: astropy/visualization/mpl_normalize.py (hunk 2 above)
Comment (larrybradley) [reply astropy/visualization/mpl_normalize.py:329]: Good catch!  This is fixed now!

File: astropy/visualization/tests/test_norm.py (hunk 3 above)
Comment (larrybradley) [reply astropy/visualization/tests/test_norm.py:312]: Fixed.  I also fixed this in the tests for the `imshow_norm` function.

File: astropy/visualization/mpl_normalize.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/visualization/mpl_normalize.py:312]: old habits die hard, don't they ? Well, moving the `*` to after `percent` is still 99% satisfactory to me, so it seems like a nice compromise !

**Comment by larrybradley** [comment]
@neutrinoceros I've added a parameterized tests for the stretches, thanks.

File: astropy/visualization/tests/test_norm.py [hunk 5]
```diff
@@ -281,31 +286,55 @@ def test_invalid_stretch(self):
...
 
 
+@pytest.mark.skipif(not HAS_PLT, reason="requires matplotlib.pyplot")
+@pytest.mark.parametrize("stretch", ["linear", "sqrt", "power", "log", "asinh", "sinh"])
+def test_simplenorm(stretch):
+    data = np.arange(25).reshape((5, 5))
+    snorm = SimpleNorm("linear", percent=99)
+    norm = snorm(data)
+    assert isinstance(norm, ImageNormalize)
+    assert_allclose(norm(data), simple_norm(data, "linear", percent=99)(data))
```
Comment (neutrinoceros) [inline astropy/visualization/tests/test_norm.py:296]: whoops
```suggestion
    snorm = SimpleNorm(stretch, percent=99)
    norm = snorm(data)
    assert isinstance(norm, ImageNormalize)
    assert_allclose(norm(data), simple_norm(data, stretch, percent=99)(data))
```

**Review by neutrinoceros — CHANGES_REQUESTED** [review]

File: astropy/visualization/tests/test_norm.py (hunk 5 above)
Comment (larrybradley) [reply astropy/visualization/tests/test_norm.py]: Fixed, thanks.  That's what I intended, but clearly I'm doing too many simultaneous tasks right now.

Comment (neutrinoceros) [reply astropy/visualization/tests/test_norm.py]: no worries, we're all in this together 😄

**Review by neutrinoceros — APPROVED** [review]

File: astropy/visualization/mpl_normalize.py [hunk 6]
```diff
@@ -212,6 +212,229 @@ def inverse(self, values, invalid=None):
...
+
+        image = np.arange(65536).reshape((256, 256))
+        snorm = SimpleNorm('sqrt', percent=98)
+        norm = snorm(image)
+        fig, ax = plt.subplots()
+        axim = ax.imshow(image, norm=norm, origin='lower')
+        fig.colorbar(axim)
+    """
+
+    def __init__(
```
Comment (neutrinoceros) [inline astropy/visualization/mpl_normalize.py:309]: just realizing now that not having type annotations here invalidates the what's new entry https://github.com/astropy/astropy/blob/main/docs/whatsnew/7.0.rst#typing-in-astropystats
I'll open a follow up PR to address this.

Comment (neutrinoceros) [reply astropy/visualization/mpl_normalize.py:309]: nevermind this isn't actually in the `stats` subpackage. My mistake.

**Pull request merged**